lib/src/socket.dart

```diff
@@ -392,211 +335,217 @@ class PhoenixSocket {
-      _logger.fine('[phoenix_socket] Heartbeat completed');
+      final heartbeatMessage = Message.heartbeat(nextRef);
+      sendMessage(heartbeatMessage);
+      _logger.fine('Heartbeat ${heartbeatMessage.ref} sent');
+      final heartbeatRef = _latestHeartbeatRef = heartbeatMessage.ref!;
+
+      _heartbeatTimeout = _scheduleHeartbeat(heartbeatRef);
+
+      await _pendingMessages[heartbeatRef]!.future;
+      _logger.fine('Heartbeat $heartbeatRef completed');
       return true;
-    } on WebSocketChannelException catch (err, stacktrace) {
-      _logger.severe(
-        '[phoenix_socket] Heartbeat message failed: WebSocketChannelException',
-        err,
-        stacktrace,
+    } on WebSocketChannelException catch (error, stackTrace) {
+      _logger.warning(
+        'Heartbeat message failed: WebSocketChannelException',
+        error,
+        stackTrace,
       );
-      _triggerChannelExceptions(PhoenixException(
-        socketError: PhoenixSocketErrorEvent(
-          error: err,
-          stacktrace: stacktrace,
+      _stateEventStreamController.add(
+        PhoenixSocketErrorEvent(
+          error: error,
+          stacktrace: stackTrace,
         ),
-      ));
-
-      return false;
-    } catch (err, stacktrace) {
-      _logger.severe(
-        '[phoenix_socket] Heartbeat message failed',
-        err,
-        stacktrace,
       );
 
+      return false;
+    } catch (error, stackTrace) {
+      _logger.warning('Heartbeat message failed', error, stackTrace);
+      _reconnect(4001, reason: 'Heartbeat timeout');
       return false;
     }
   }
 
-  void _triggerChannelExceptions(PhoenixException exception) {
-    _logger.fine(
-      () => 'Trigger channel exceptions on ${channels.length} channels',
-    );
-    for (final channel in channels.values) {
-      _logger.finer(
-        () => 'Trigger channel exceptions on ${channel.topic}',
+  Timer _scheduleHeartbeat([String? previousHeartbeat]) {
+    return Timer(_options.heartbeat, () {
+      if (previousHeartbeat != null) {
+        final completer = _pendingMessages.remove(previousHeartbeat);
+        if (completer != null && !completer.isCompleted) {
+          completer.completeError(
+            TimeoutException(
+              'Heartbeat $previousHeartbeat not completed before sending new one',
+            ),
+            StackTrace.current,
+          );
+          return;
+        }
+      }
+
+      _sendHeartbeat();
+    });
+  }
+
+  void _cancelHeartbeat() {
+    if (_heartbeatTimeout != null) {
+      _heartbeatTimeout?.cancel();
+      _heartbeatTimeout = null;
+    }
+    if (_latestHeartbeatRef != null) {
+      _pendingMessages.remove(_latestHeartbeatRef)?.complete(
+            Message.heartbeat(_latestHeartbeatRef!), // doesn't matter
+          );
+      _latestHeartbeatRef = null;
+    }
+  }
+
+  bool _shouldPipeMessage(String message) {
+    final currentSocketState = _socketStateStream.valueOrNull;
```

Comment:
Oh, forgot to leave a comment here. This could become a situation for a "black hole" where a user isn't receiving events but "we don't know why."

I can't say this will be a problem, but I wonder if we should throw an exception or log something here for the first version to ensure we aren't dropping anything we might want to keep.